server/src/org/apache/cloudstack/affinity/AffinityGroupServiceImpl.java

```diff
@@ -205,6 +208,20 @@ public AffinityGroupVO doInTransaction(TransactionStatus status) {
         });
     }
 
+    /**
+     * If the account is null (domainLevel is true), then returns the domain id passed as a
+     * parameter; otherwise (domainLevel is false) it returns the domain id from the owner account.
+     *
+     * @note: this method fixes a critical bug. More details in JIRA ticket CLOUDSTACK-9432.
+     */
+    protected Long getDomainIdBasedOnDomainLevel(final Account owner, boolean domainLevel, Long domainId) {
+        Long domainIdBasedOnDomainLevel = owner.getDomainId();
+        if (domainLevel) {
+            domainIdBasedOnDomainLevel = domainId;
+        }
+        return domainIdBasedOnDomainLevel;
+    }
+
```

Comment:
@rafaelweingartner thanks for the review. The code now is in a method with tests and documentation.